blockstore/api.go

```diff
@@ -31,31 +40,44 @@ func (a *apiBlockstore) DeleteBlock(context.Context, cid.Cid) error {
 }
 
 func (a *apiBlockstore) Has(ctx context.Context, c cid.Cid) (bool, error) {
+	if a.cache.Contains(c) {
+		return true, nil
+	}
 	return a.api.ChainHasObj(ctx, c)
 }
 
 func (a *apiBlockstore) Get(ctx context.Context, c cid.Cid) (blocks.Block, error) {
+	if bb, ok := a.cache.Get(c); ok {
+		return blocks.NewBlockWithCid(bb, c)
+	}
 	bb, err := a.api.ChainReadObj(ctx, c)
 	if err != nil {
 		return nil, err
 	}
+	a.cache.Add(c, bb)
 	return blocks.NewBlockWithCid(bb, c)
 }
 
 func (a *apiBlockstore) GetSize(ctx context.Context, c cid.Cid) (int, error) {
+	if bb, ok := a.cache.Get(c); ok {
+		return len(bb), nil
+	}
 	bb, err := a.api.ChainReadObj(ctx, c)
 	if err != nil {
 		return 0, err
 	}
+	a.cache.Add(c, bb)
 	return len(bb), nil
 }
 
 func (a *apiBlockstore) Put(ctx context.Context, block blocks.Block) error {
+	a.cache.Add(block.Cid(), block.RawData())
```

Comment:
Here we probably want to update the recentness? If not, ditto re `ContainsOrAdd`.